File: git_push_all.py
import os
import sys
import subprocess

# Function to check and install GitPython
def install_gitpython():
    try:
        import git  # Try importing GitPython
    except ImportError:
        print("🔄 GitPython is not installed. Installing it now...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", "gitpython"])
        import git  # Import again after installation

# Ensure GitPython is installed before running the script
install_gitpython()

import git  # Now it's safe to import GitPython

# Get the current directory (assumes the script is inside the project folder)
project_path = os.getcwd()

print(f"🔄 Detected project directory: {project_path}")

# Check if it's a valid Git repository
if not os.path.exists(os.path.join(project_path, ".git")):
    print("❌ This directory is not a Git repository. Please initialize it first using 'git init'.")
    sys.exit(1)

try:
    # Load the Git repository
    repo = git.Repo(project_path)

    # Show git status
    print("🔍 Checking for changes...")
    print(repo.git.status())

    # Stage all changes
    print("📌 Staging all changes...")
    repo.git.add(all=True)

    # Ask for commit message
    commit_message = input("✍️  Enter your commit message: ").strip()
    if not commit_message:
        print("❌ Commit message cannot be empty. Please run the script again with a valid message.")
        sys.exit(1)

    # Commit changes
    repo.git.commit(m=commit_message)

    # Push changes to GitHub
    print("📤 Pushing changes to GitHub...")
    repo.git.push("origin", "main")

    print("✅ Push completed successfully!")

except Exception as e:
    print(f"❌ Push failed! Error: {e}")
    print("Try running 'git pull origin main --rebase' first.")
    sys.exit(1)

File: git_push_all.cpp
#include <iostream>
#include <cstdlib>   // For system() and OS detection
#include <git2.h>    // libgit2 header
#include <string>

using namespace std;

// Function to check and install libgit2 automatically
void installLibgit2() {
#ifdef _WIN32
    cout << "🔍 Checking for libgit2 on Windows..." << endl;
    system("where libgit2 || (echo Installing libgit2... & vcpkg install libgit2)");
#elif __APPLE__
    cout << "🔍 Checking for libgit2 on macOS..." << endl;
    system("brew list libgit2 || (echo Installing libgit2... && brew install libgit2)");
#elif __linux__
    cout << "🔍 Checking for libgit2 on Linux..." << endl;
    system("dpkg -s libgit2-dev || (echo Installing libgit2... && sudo apt update && sudo apt install -y libgit2-dev)");
#else
    cout << "❌ Unsupported OS. Install libgit2 manually." << endl;
    exit(1);
#endif
}

// Function to commit and push changes
void gitCommitAndPush() {
    git_libgit2_init();  // Initialize libgit2

    git_repository* repo = nullptr;
    if (git_repository_open(&repo, ".")) {
        cerr << "❌ Error: Not a valid Git repository!" << endl;
        exit(1);
    }

    cout << "✅ Git repository detected!" << endl;

    // Stage all changes
    system("git add .");

    // Get commit message from user
    string commitMessage;
    cout << "✍️  Enter your commit message: ";
    getline(cin, commitMessage);

    if (commitMessage.empty()) {
        cerr << "❌ Commit message cannot be empty!" << endl;
        exit(1);
    }

    // Commit changes
    string commitCmd = "git commit -m \"" + commitMessage + "\"";
    system(commitCmd.c_str());

    // Push to GitHub
    cout << "📤 Pushing changes to GitHub..." << endl;
    system("git push origin main");

    cout << "✅ Push completed successfully!" << endl;

    git_repository_free(repo);
    git_libgit2_shutdown();  // Clean up libgit2
}

int main() {
    cout << "🚀 Starting Git Push Automation..." << endl;

    // Step 1: Install libgit2 if missing
    installLibgit2();

    // Step 2: Run Git operations
    gitCommitAndPush();

    return 0;
}
